flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -1505,22 +1505,27 @@ class ScalarExprLowering {
     return std::visit([&](const auto &x) { return genval(x); }, dref.u);
   }
 
-  // Helper function to turn the left-recursive Component structure into a list
-  // that does not contain allocatable or pointer components other than the last
-  // one.
-  // Returns the object used as the base coordinate for the component chain.
+  // Helper function to turn the Component structure into a list of nested
+  // components, ordered from largest/leftmost to smallest/rightmost:
+  //  - where only the smallest/rightmost item may be allocatable or a pointer
+  //    (nested allocatable/pointer components require nested coordinate_of ops)
+  //  - that does not contain any parent components
+  //    (the front end places parent components directly in the object)
+  // Return the object used as the base coordinate for the component chain.
   static Fortran::evaluate::DataRef const *
   reverseComponents(const Fortran::evaluate::Component &cmpt,
-                    std::list<const Fortran::evaluate::Component *> &list) {
-    list.push_front(&cmpt);
+                    std::list<const Fortran::evaluate::Component *> &list,
+                    bool isParentComponent = false) {
```

Comment:
I don't think you need an extra flag here to indicate something about the state of `cmpt`. Just test for it.

```c++
  static Fortran::evaluate::DataRef const *
  reverseComponents(const Fortran::evaluate::Component &cmpt,
                    std::list<const Fortran::evaluate::Component *> &list) {
    if (!cmpt.GetLastSymbol().test(
            Fortran::semantics::Symbol::Flag::ParentComp))
      list.push_front(&cmpt);
    return std::visit(
        Fortran::common::visitors{
            [&](const Fortran::evaluate::Component &x) {
              const Fortran::semantics::Symbol &sym = x.GetLastSymbol();
              if (Fortran::semantics::IsAllocatableOrPointer(sym))
                return &cmpt.base();
              return reverseComponents(x, list);
            },
            [&](auto &) { return &cmpt.base(); },
        },
        cmpt.base().u);
  }
```

Reply:
Yes, that is better; thanks.